### ISSUE TITLE
FR: `git commit --verbose` equivalent for `jj describe`

### ISSUE THREAD

**Issue by sullyj3** [jj-vcs/jj#1946]
Often when writing a commit message, it's useful to be able to see the diff at a glance. Git has `git commit --verbose`, which displays the diff inline in your editor while you're writing the message. This saves having to try to remember what changes you made. 
I personally use this all the time. I have 
```
alias gca='git commit --all --verbose'
```
which I use for committing almost exclusively.

It would be cool if `jj` supported this as well! I'm not wedded to the name `--verbose`, which I think could be clearer (and in fact looking at the help it conflicts with an existing flag with different functionality). Maybe `--show-diff`?

**Comment by ilyagr** [comment]
This could apply to quite a few commands (`describe`, `split`, `squash`, etc)

**Comment by martinvonz** [comment]
Maybe we can have a configurable template for producing that. We currently include something like this:

```
JJ: This commit contains the following changes:
JJ:     M lib/src/working_copy.rs
JJ:     M lib/tests/test_working_copy.rs
```

Maybe that could be replaced by a template like `"This commit contains the following changes:\n" ++ indent("    ", diffsummary)`, and you could replace the `diffsummary` by `diff` (or `gitdiff`, perhaps) to get the full diff.

We don't yet have support for including a diff in templates, so we'd need to implement that first.

#1354 is slightly related.

**Comment by 9999years** [comment]
I went looking for this feature, remembered `jj` has templates, and was a little surprised that `jj describe` didn't have a `--template` option for this yet, so I think templates would be a natural place to implement this. (Although I suspect this is a common enough use-case that a dedicated flag might be worthwhile, even if it's just an alias for `--template=commit-with-diff` or something.)

**Comment by Cretezy** [comment]
Maybe `jj describe --diff` should be the command? `jj split --diff` and `jj squash --diff` too.

**Comment by Cretezy** [comment]
Looking for consensus before implementing this:

1. Should we use `--show-diff` for `diff`/`squash`/`split`?
2. Should we instead use a template feature for this?
  1. If using template, how will this be used? Will jj ship with a built-in template for this?

**Comment by yuja** [comment]
> 1. Should we use `--show-diff` for `diff`/`squash`/`split`?

I think config knob is better. It was discussed previously in
https://github.com/martinvonz/jj/pull/3470#issuecomment-2053014846

> 2. Should we instead use a template feature for this?

Template is preferred because it can solve other problems (such as #1354 and #3385), but it's not easy. I think adding a diff formatting option is also good as a short-term workaround.

> 3. If using template, how will this be used? Will jj ship with a built-in template for this?

Yeah. Perhaps, the default `ui.default-description` will be a template like
```
concat(
  description,
  "\n",
  "JJ: This commit contains the following changes:\n",
  indent("JJ:     ", format_diff_in_commit_description(diff)),
)
```
and `format_diff_in_commit_description()` will be customized by user.

**Comment by yuja** [comment]
> Template is preferred because it can solve other problems (such as #1354 and #3385), but it's not easy.

FWIW, I'm working on `commit.diff([paths])` template function, and I think a basic version can be added soon.

**Comment by lukerandall** [comment]
For those who stumble upon this; the nearest equivalent (since #4153) is to add something like this to your config file:

```
[templates]
draft_commit_description = '''
concat(
  description,
  surround(
    "\nJJ: This commit contains the following changes:\n", "",
    indent("JJ:     ", diff.stat(72)),
  ),
  surround(
    "\nJJ: Diff:\n", "",
    indent("JJ:     ", diff.git(72)),
  ),
)
'''
```

or if you don't want it all the time add an alias using `--config-toml` (in this example I use a template-alias to make the alias easier to read):

```
[aliases]
diffscribe = [
  "describe",
  "--config-toml=templates.draft_commit_description=\"commit_description_with_diff\"",
]

[template-aliases]
commit_description_with_diff = '''
concat(
  description,
  surround(
    "\nJJ: This commit contains the following changes:\n", "",
    indent("JJ:     ", diff.stat(72)),
  ),
  surround(
    "\nJJ: Diff:\n", "",
    indent("JJ:     ", diff.git(72)),
  ),
)
'''


The most notable deficiency is that you don't get diff highlighting in your text editor like you would with a git commit diff because of the prefixed `JJ:`.

**Comment by tbodt** [comment]
> ```
>  surround(
>    "\nJJ: Diff:\n", "",
>    indent("JJ:     ", diff.git(72)),
>  ),
> ```

Important note - you almost certainly don't want `diff.git(72)`, because the 72 becomes the number of lines of diff context! `diff.git()` appears to default to 3 lines of context which is fine.

**Comment by bryceberger** [comment]
Since #5155, you can now do something like:
```toml
[template-aliases]
commit_description_verbose = '''
concat(
  description,
  "\n",
  "JJ: This commit contains the following changes:\n",
  indent("JJ:    ", diff.stat(72)),
  "JJ: ignore-rest\n",
  diff.git(),
)
'''

[aliases]
dv = ["--config=templates.draft_commit_description=commit_description_verbose", "describe"]
```

**Comment by yuja** [comment]
Thanks. I think we can now close this issue?

**Comment by oyarsa** [comment]
This doesn't highlight changes in Vim like `git commit --verbose` does (additions are green, and removals are red). Is it possible to do that?

**Comment by Samasaur1** [comment]
I got that working in Neovim with a jjdescription tree-sitter plugin, so it's definitely possible

**Comment by oyarsa** [comment]
> I got that working in Neovim with a jjdescription tree-sitter plugin, so it's definitely possible

Could you give more details?

**Comment by pylbrecht** [comment]
This is just based on a few minutes of research, but you might be able to add [this jj-description grammar](https://github.com/kareigu/tree-sitter-jjdescription) by following [treesitter's instuctions on adding parsers](https://github.com/nvim-treesitter/nvim-treesitter?tab=readme-ov-file#adding-parsers).

**Comment by pylbrecht** [comment]
Ok, here is a quick and dirty solution for [lazy.nvim](https://lazy.folke.io/).

1. Add the jjdescription parser to nvim-treesitter.

    Add this to your [lazy plugin spec](https://lazy.folke.io/spec#spec-setup) (using lazy.nvim's [structured setup](https://lazy.folke.io/installation)):
    
    ```lua
    {
     'nvim-treesitter/nvim-treesitter',
      config = function()
        local parser_config = require "nvim-treesitter.parsers".get_parser_configs()
        parser_config.jjdescription = {
          install_info = {
            url = "https://github.com/kareigu/tree-sitter-jjdescription",
            files = { "src/parser.c" },
            branch = "dev",
          },
          filetype = "jj",
        }
        vim.treesitter.language.register('jjdescription', 'jj')
      end,
    }
    ```
    *(See https://github.com/nvim-treesitter/nvim-treesitter?tab=readme-ov-file#adding-parsers for reference.)*
    
3. Manually add query files to your local nvim-treesitter installation.
    
    I did this by manually copying [tree-sitter-jjdescription’s query files](https://github.com/kareigu/tree-sitter-jjdescription/tree/d09205b52b5a0165b588a793e366c1116468d86f/queries) to `~/.local/share/nvim/lazy/nvim-treesitter/queries/jjdescription/`.
    
    *(See https://github.com/nvim-treesitter/nvim-treesitter?tab=readme-ov-file#adding-queries for reference.)*
    
4. Start `nvim` and `:TSInstall jjdescription`

Using the config from https://github.com/jj-vcs/jj/issues/1946#issuecomment-2561045057 you should see a colorful diff (assuming you have changes in your working copy) when running

```bash
$ jj dv
```
![image](https://github.com/user-attachments/assets/16c7d9b4-51fa-4613-8b48-44642aa8c72f)

---

There is still room for improvement (e.g. automating the manual copying of query files), but I hope this serves as a good starting point anyway.

**Comment by Samasaur1** [comment]
Sorry, was out and about. @pylbrecht got it exactly right. My config is with Nix, so it's probably harder to apply, but I can post the diff of my config if you'd like

**Comment by Samasaur1** [comment]
One other thing I did was setting:
```toml
[templates]
draft_commit_description = '''
concat(
  description,
  surround(
    "\nJJ: This commit contains the following changes:\n", "",
    indent("JJ:     ", diff.summary()),
  ),
  surround(
    "\nJJ: This commit contains the following changes:\n", "",
    indent("JJ:     ", diff.stat(72)),
  ),
  "\n",
  "JJ: ignore-rest\n",
  diff.git(),
)
'''
```
so that normal `jj commit` and `jj describe` use the verbose form. That's obviously personal preference though.

Also worth noting that if you use `diff.summary()` instead of `diff.stat(72)`, the tree-sitter plugin will highlight that as well

**Comment by stacyharper** [comment]
For those using Kakoune, here the simplest but working filetype:

https://git.sr.ht/~stacyharper/dotfiles/tree/master/item/.config/kak/autoload/user/jj.kak

**Comment by gpanders** [comment]
FYI both Vim and Neovim support jj description syntax highlighting, including the diff syntax now https://github.com/vim/vim/pull/16364 No tree-sitter or plugins required.

**Comment by Samasaur1** [comment]
> FYI both Vim and Neovim support jj description syntax highlighting, including the diff syntax now [vim/vim#16364](https://github.com/vim/vim/pull/16364) No tree-sitter or plugins required.

Am I correct that this needs latest Neovim master/the upcoming v0.10.4?

**Comment by gpanders** [comment]
> Am I correct that this needs latest Neovim master/the upcoming v0.10.4?

Yes

**Comment by sersorrel** [comment]
for the sake of anyone else finding this issue: neovim 0.10.4 was released without this (at a guess, https://github.com/neovim/neovim/pull/31840 was merged after 0.10 branchoff?), afaict it will only land in 0.11 (going by github milestones, this will arrive in mid-march https://github.com/neovim/neovim/milestone/41).
